Implemented zpool scrub pause/resume

alek-p opened the pull request:
Currently, there is no way to pause a scrub. Pausing may
be useful when the pool is busy with other I/O to preserve
bandwidth.

### Description
This patch adds the ability to pause and resume scrubbing.
This is achieved by maintaining a persistent on-disk scrub state.
While the state is 'paused' we do not scrub any more blocks.
We do however perform regular scan housekeeping such as
freeing async destroyed and deadlist blocks while paused.

### Motivation and Context
Scrub pausing can be an I/O intensive operation and people have been asking for the ability to pause a scrub for a while. This allows one to preserve scrub progress while freeing up bandwidth for other I/O.

### How Has This Been Tested?
Unit testing and zfs-tests.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux code style requirements.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

alek-p commented:
looks like I missed the makefile update for `zpool_import_scrub_pause_001_pos.ksh`. I'll fix that tonight

loli10K commented:
> This patch also adds a -p option to zpool import which allows one to pause scrub instead of resuming it on import.

This seems a suitable fix for https://github.com/zfsonlinux/zfs/issues/3658

alek-p commented:
Looks like the problem was something else:
```
tar: zfs-0.7.0/tests/zfs-tests/tests/functional/cli_root/zpool_import/zpool_import_scrub_pause_001_pos.ksh: file name is too long (max 99); not dumped
tar: Exiting with failure status due to previous errors
```
I've changed the filename, let's try again...

loli10K commented on module/zfs/zfs_ioctl.c:
```diff
@@ -1683,7 +1684,12 @@ zfs_ioc_pool_scan(zfs_cmd_t *zc)
 		return (error);
 
-	if (zc->zc_cookie == POOL_SCAN_NONE)
+	ASSERT(zc->zc_flags >= POOL_SCRUB_NORMAL &&
```
Can we be a little more tolerant here and `return (SET_ERROR(EINVAL))` instead of ASSERTing `zc_flags` value?

```
root@debian-8-zfs:~# stap -g -v -d `which zfs` --ldd \
>    -e 'probe process("/lib*/libzfs.so*").function("zpool_scan").call { $cmd = 0xff; }' \
>    -c "zpool scrub -p $POOLNAME"
Pass 1: parsed user script and 113 library scripts using 94688virt/39056res/5008shr/34604data kb, in 390usr/100sys/1727real ms.
Pass 2: analyzed script: 1 probe, 1 function, 0 embeds, 0 globals using 95856virt/41876res/6536shr/35772data kb, in 20usr/0sys/118real ms.
Pass 3: using cached /root/.systemtap/cache/dd/stap_dd9957a3d2f294127e57b1abbc2c4bac_2280.c
Pass 4: using cached /root/.systemtap/cache/dd/stap_dd9957a3d2f294127e57b1abbc2c4bac_2280.ko
Pass 5: starting run.
[   72.776788] Kprobes globally unoptimized
[   72.929149] VERIFY(zc->zc_flags >= POOL_SCRUB_NORMAL && zc->zc_flags < POOL_SCRUB_FLAGS_END) failed
[   72.930278] PANIC at zfs_ioctl.c:1688:zfs_ioc_pool_scan()
[   72.932426] Showing stack for process 2024
[   72.935176] CPU: 7 PID: 2024 Comm: zpool Tainted: P           OE   4.6.4 #3
[   72.935758] Hardware name: Bochs Bochs, BIOS Bochs 01/01/2011
[   72.936396]  0000000000000286 0000000047860cbc ffff8800758a7c20 ffffffff81491063
[   72.938873]  ffffffffc0450268Message from sy 0000000000000698 ffff8800758a7c30 ffffffffc005bd04
[   72.942586]  ffff8800758a7db8 ffffffffc005bdcf 0000000000000007 ffff880000000028
[   72.943316] Call Trace:
[   72.943551]  [<ffffffff81491063>] dump_stack+0x63/0x90
[   72.943998]  [<ffffffffc005bd04>] spl_dumpstack+0x44/0x50 [spl]
[   72.944501]  [<ffffffffc005bdcf>] spl_panic+0xbf/0xf0 [spl]
[   72.946604]  [<ffffffff81200200>] ? create_object+0x170/0x2f0
[   72.948138]  [<ffffffff81812452>] ? mutex_lock+0x12/0x30
[   72.950116]  [<ffffffffc02e6433>] ? refcount_add_many+0x73/0xf0 [zfs]
[   72.950847]  [<ffffffffc02e64c6>] ? zfs_refcount_add+0x16/0x20 [zfs]
[   72.951437]  [<ffffffffc030063b>] ? spa_open_ref+0x3b/0x90 [zfs]
[   72.952031]  [<ffffffffc02f7b48>] ? spa_open_common+0x2b8/0x4b0 [zfs]
[   72.954031]  [<ffffffffc033f916>] zfs_ioc_pool_scan+0xc6/0xe0 [zfs]
[   72.956447]  [<ffffffffc0346525>] zfsdev_ioctl+0x635/0x750 [zfs]
[   72.958340]  [<ffffffff8117a934>] ? __create_xol_area+0x1a4/0x210
[   72.960631]  [<ffffffff810636fe>] ? arch_uprobe_post_xol+0x6e/0x100
[   72.962671]  [<ffffffff8121ac4d>] do_vfs_ioctl+0x9d/0x5b0
[   72.964428]  [<ffffffffc01e6617>] ? utrace_report_syscall_entry+0x117/0x130 [stap_dd9957a3d2f294127e57b1abbc2c4bac_2021]
[   72.968223]  [<ffffffff8121b1d9>] SyS_ioctl+0x79/0x90
[   72.969048]  [<ffffffff81003cf9>] do_syscall_64+0x89/0x180
[   72.970008]  [<ffffffff81814425>] entry_SYSCALL64_slow_path+0x25/0x25
```

alek-p replied:
sure, I made the change

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -5864,8 +5884,17 @@ print_scan_status(pool_scan_stat_t *ps)
+			size_t len = strlen(ctime(&pause));
+			char cbuf[len];
+			struct tm *p = localtime(&pause);
+			(void) strftime(cbuf, len, "%a %b %e %T %Y", p);
```
Using `strlen(ctime())` as a way to calculate the maximum buffer size for `strftime()` is a little dangerous since the required buffer size depends on the pattern format specifier.  It would be simpler to safer to declare this buffer as say 32 bytes.

behlendorf commented on include/sys/dsl_scan.h:
```diff
@@ -107,7 +110,9 @@ typedef enum dsl_scan_flags {
-	boolean_t scn_pausing;
+	uint8_t scn_scrub_state;
+	uint64_t scn_scrub_pause_time;
+	boolean_t scn_suspending;
```
It would be better to move `boolean_t scn_suspending` before `uint64_t scn_scrub_pause_time` to make sure this structure is nicely packed.

behlendorf commented on lib/libzfs/libzfs_util.c:
```diff
@@ -246,6 +247,9 @@ libzfs_error_description(libzfs_handle_t *hdl)
 		    "into a new one"));
+	case EZFS_SCRUB_PAUSED:
+		return (dgettext(TEXT_DOMAIN, "scrub currently paused; "
+		    "use 'zpool scrub' to resume scrubbing"));
```
Exceeds 80 character limit.

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -5890,15 +5918,21 @@ print_scan_status(pool_scan_stat_t *ps)
+		    examined_buf, total_buf, rate_buf);
+		if (hours_left < (30 * 24)) {
+			(void) printf(gettext(", %lluh%um to go\n"),
+			    (u_longlong_t)hours_left, (uint_t)(mins_left % 60));
```
I might be mistaken about this, but it looks like the calculated rate may include the time paused in the calculation.  If that's the case, we'll want to exclude it to keep the `zpool status` output useful.

ahrens commented on include/sys/dsl_scan.h:
```diff
@@ -107,7 +110,9 @@ typedef enum dsl_scan_flags {
 
-	boolean_t scn_pausing;
+	boolean_t scn_suspending;
+	uint8_t scn_scrub_state;
```
should this be an enum?  this struct is in memory and there aren't many instances of it, so using a larger type is fine.

ahrens commented on man/man8/zpool.8:
```diff
@@ -2045,6 +2063,17 @@ Because scrubbing and resilvering are \fBI/O\fR-intensive operations, \fBZFS\fR
+Pause scrubbing. Scrub progress is periodically synced to disk so if the system
+is restarted or pool is exported during a paused scrub, the scrub will resume
+from the place where it was last checkpointed to disk. To resume a paused scrub
+issue \fBzpool scrub\fR again.
```
I don't see where the paused-ness is persisted to disk.  So when the pool is opened/imported, it will not be paused, even if it was paused before export/reboot (unless you do `zpool import -p`)?  We should at least document that confusing behavior.  But I would much prefer to fix it before integrating this.

I think you could either:
1. add a feature flag and introduce a new dsl_scan_state_t; or 
2. add a new dsl_scan_flags_t, which would be ignored on older software (so if the pool is imported/opened with older software, the scrub would not be paused)

behlendorf replied:
I'd be nice to avoid the feature flag and reserve a `dsl_scan_flags_t` bit for this to avoid potential GRUB compatibility issues.

behlendorf commented:
@alek-p FYI a rebase on master will resolve the [memory leak](http://build.zfsonlinux.org/builders/Ubuntu%2016.04%20x86_64%20Kmemleak%20%28TEST%29/builds/1935/steps/shell_9/logs/kmemleak) detected by the kmemleak builder.

alek-p commented:
thanks for the reviews, I should be able to address these comments in the next couple of days

alek-p commented:
In the updated patch I've shortened the error messages as much as I could without making them grammatically too ugly. Keep in mind some of the existing ones are also over 80 chars.
I've also modified the patch to take into account the time spent paused so that the rate isn't affected by paused time.
Finally, the scrub paused state is now persistent on disk using option 2 (new dsl_scan_flags_t) suggested by Matt so there is no new feature flag.

loli10K commented on module/zfs/dsl_scan.c:
```diff
@@ -141,9 +142,9 @@ dsl_scan_init(dsl_pool_t *dp, uint64_t txg)
 		 */
 		if (err == EOVERFLOW) {
```
I wonder if adding a new value to `dsl_scan_phys_t` (and bumping `SCAN_PHYS_NUMINTS` to 25) will prevent us from detecting errata 2094 in the first place, since `zap_lookup` will not EOVERFLOW if the pool is affected.

alek-p replied:
I think I can avoid modifying the on-disk format (`dsl_scan_phys_t`) here, as `scn_scrub_pause_time` doesn't need to be persistent and that will actually make the code simpler.

However, you do bring up an interesting point - errata 2094 locks us into the current on-disk format for dsl_scan_phys_t...

ahrens replied:
it locks you into the current *size* of the dsl_scan_phys_t, but adding the new flag should be OK.

ahrens commented on include/sys/dsl_scan.h:
```diff
@@ -55,6 +56,7 @@ typedef struct dsl_scan_phys {
 	uint64_t scn_cur_max_txg;
 	uint64_t scn_start_time;
 	uint64_t scn_end_time;
+	uint64_t scn_scrub_pause_time; /* when scrub was paused */
```
regardless of the other errata, you definitely can't change the offset of the following fields without adding a feature flag and managing the transition to the new on-disk format.

alek-p replied:
yea I realized that after @loli10K's comment about errata. What I'm going to have to do avoid changing on disk format is have the pause time reset every time the pool is imported in a paused state which I think is ok.

alek-p commented:
With the latest patch scrub-pause persistence is now achieved without any on-disk format changes.
There are caveats that follow from that:
1) If exported pool had scrub paused the scrub pause time as seen in zpool status output is initialised on import.
2) Scrub completion time includes time spent paused. It also includes time spent exported just like before this patch.

ahrens commented:
@alek-p Thanks!  Those caveats sound reasonable to me.

behlendorf commented:
@alek-p thanks, that sounds reasonable.

behlendorf commented on module/zfs/zfs_ioctl.c:
```diff
@@ -1683,7 +1684,13 @@ zfs_ioc_pool_scan(zfs_cmd_t *zc)
 		return (error);
 
-	if (zc->zc_cookie == POOL_SCAN_NONE)
+	if (zc->zc_flags < POOL_SCRUB_NORMAL ||
```
This can never be true since `zc_flags` is unsigned and `POOL_SCRUB_NORMAL = 0`.  Let's drop this first check, it's going to upset static code analysis utilities.

behlendorf commented on tests/zfs-tests/include/libtest.shlib:
```diff
@@ -2059,6 +2059,12 @@ function is_pool_scrub_stopped #pool
+{
+	check_pool_status "$1" "scan" "scrub paused since "
+	return $?
+}
```
You'll want to update the comment above `is_pool_resilvering` to include this additional helper function.  Including changing the 5->6.

alek-p replied:
thanks, I've made the edits

behlendorf commented on tests/zfs-tests/tests/functional/cli_root/zpool_import/zpool_import_scrub_pause.ksh:
```diff
@@ -0,0 +1,81 @@
+
+#
+# Copyright (c) 2012, 2016 by Delphix. All rights reserved.
+# Copyright (c) 2017 Datto Inc.
```
You can group all of the copyright lines together.

```
# Copyright 2007 Sun Microsystems, Inc.  All rights reserved.
# Copyright (c) 2012, 2016 by Delphix. All rights reserved.
# Copyright (c) 2017 Datto Inc.
# Use is subject to license terms.
```
￼

￼

ahrens commented on module/zfs/dsl_scan.c:
```diff
@@ -403,6 +423,56 @@ dsl_scan_cancel(dsl_pool_t *dp)
+
+	/* can't pause a scrub when there is no in-progress scrub */
+	if (cmd == POOL_SCRUB_PAUSE && !dsl_scan_scrubbing(dp))
+		return (SET_ERROR(ENOENT));
```
I think this function would be easier to follow if it was overall structured like:

```
if (cmd == PAUSE) {
  // do checks
  // set fields
} else {
  // ASSERT(cmd == NORMAL)
  if (paused_scrub()) {
    // set fields
  }
}
```

alek-p replied:
sure

ahrens commented on module/zfs/dsl_scan.c:
```diff
@@ -1686,6 +1756,12 @@ dsl_scan_sync(dsl_pool_t *dp, dmu_tx_t *tx)
 
+	if (dsl_scan_is_paused_scrub(scn)) {
+		zfs_dbgmsg("bailing on scrub since it's paused, sync txg %llu",
+		    (longlong_t)tx->tx_txg);
```
This message might be a bit excessive, since it will be printed every txg, even though we aren't doing anything.  My general philosophy around zfs_dbgmsg is that we should use it when something exceptional happens, when we make roughly once-per-txg on-disk changes, and roughly once-per-txg when we make progress on background tasks.  But simply saying "the system continues to be in this normal state" every txg seems like a waste of log space.

alek-p replied:
ok I'll drop the message

ahrens commented on module/zfs/txg.c:
```diff
@@ -500,7 +501,8 @@ txg_sync_thread(dsl_pool_t *dp)
 		timer = (delta >= timeout ? 0 : timeout - delta);
-		while (!dsl_scan_active(dp->dp_scan) &&
+		while ((!dsl_scan_active(dp->dp_scan) ||
+		    dsl_scan_is_paused_scrub(dp->dp_scan)) &&
```
can we instead change dsl_scan_active() to return FALSE when it's paused?

alek-p replied:
sure

ahrens commented on module/zfs/dsl_scan.c:
```diff
@@ -403,6 +423,56 @@ dsl_scan_cancel(dsl_pool_t *dp)
+
+	if (cmd == POOL_SCRUB_PAUSE) {
+		spa->spa_scan_pass_scrub_pause = gethrestime_sec();
+		scn->scn_phys.scn_flags |= DSF_SCRUB_PAUSED;
```
Is it OK to do this from open context (I'm guessing no)?  When will it be persisted to disk (I'm guessing never / by luck, since you are not calling dsl_scan_state_sync(), which can't be called from open context anyway)?

I would suggest that you do this from a synctask instead.

alek-p replied:
I belive dsl_scan_sync_state() is called every txg by way of spa_sync() -> dsl_scan_sync() -> dsl_scan_sync_state().

ahrens replied:
It can be called by dsl_scan_sync(), but I don't think it is called if the scrub is paused.  dsl_scan_sync() will return before getting to the end, where it calls dsl_scan_sync_state().

alek-p replied:
Ok, I'll change it this to be done from a sync task

ahrens commented on module/zfs/dsl_scan.c:
```diff
@@ -403,6 +423,56 @@ dsl_scan_cancel(dsl_pool_t *dp)
+		spa->spa_scan_pass_scrub_spent_paused +=
+		    gethrestime_sec() - spa->spa_scan_pass_scrub_pause;
+		spa->spa_scan_pass_scrub_pause = 0;
+		scn->scn_phys.scn_flags &= ~DSF_SCRUB_PAUSED;
```
same issue here with modifying scn_phys.

ahrens commented on tests/zfs-tests/tests/functional/cli_root/zpool_import/zpool_import_scrub_pause.ksh:
```diff
@@ -0,0 +1,78 @@
+	log_must mkfile $FILE_SIZE $VDEV0 $VDEV1
+}
+
+log_assert "Importing a pool with the pause scrub option works as expected"
```
this doesn't seem accurate

behlendorf commented on cmd/zpool/zpool_main.c:
```diff
@@ -2389,6 +2392,9 @@ zpool_do_import(int argc, char **argv)
 			break;
+		case 'p':
+			flags |= ZFS_IMPORT_SCRUB_PAUSE;
+			break;
```
We've had cases in the past where this would have been useful.  For example issue #3658, where due to some unfortunate hardware/drivers the pool couldn't be imported and the scrub canceled quickly enough to avoid other problems.  This might also be a useful feature to aid with pool recovery scenarios.

ahrens replied:
For pool recovery, importing readonly would be a better option.

I get that "import + cancel/pause scrub" could be useful in a few specific scenarios, but from a user interface design it seems arbitrary.  There are a lot of other specific scenarios where other combo options could be helpful.  e.g. what if I'm resilvering and for some reason the hardware doesn't like that?  What if I'm doing a background destroy?  What if one of the disks is sort of busted and I want to do a "import + offline <dev>"?

behlendorf replied:
For pool recovery, importing readonly is absolutely the best option to start with.  In fact, if scrubs were automatically paused for readonly pools that would address my use case.

I see your point about this option being somewhat arbitrary.  I'm not opposed to dropping it particularly if we made the suggested readonly tweak.  But perhaps @alek-p  has another use case in mind?

ahrens replied:
Scrubs have always been (silently) paused on readonly pools.  Scrub happens from spa_sync(), and requires modifying the on-disk format (the scn_queue_obj), so it isn't possible to scrub a readonly pool.

alek-p replied:
I'm fine with dropping the `import -p` option. My use case was the same - to address #3658

alek-p commented on include/sys/dsl_scan.h:
```diff
@@ -70,6 +71,7 @@ typedef struct dsl_scan_phys {
 
 typedef enum dsl_scan_flags {
 	DSF_VISIT_DS_AGAIN = 1<<0,
+	DSF_SCRUB_PAUSED = 1<<1,
```
I can just do a full upstream of scrub-pause to OpenZFS. Unless we want soak time in ZoL first?

ahrens replied:
Full upstream would be great!

behlendorf replied:
Concurrent PRs against OpenZFS and ZoL would be great.

alek-p replied:
done, see https://github.com/openzfs/openzfs/pull/407

alek-p commented on man/man8/zpool.8:
```diff
@@ -2044,6 +2062,17 @@ Because scrubbing and resilvering are \fBI/O\fR-intensive operations, \fBZFS\fR
+\fB\fB-p\fR\fR
+.ad
+.RS 6n
+Pause scrubbing. Scrub progress is periodically synced to disk so if the system
```
I can add a sentence at the end: "Note, resilver can't be paused."
However, this is documenting the scrub command so I'm not sure resilver notes belong here?

ahrens replied:
That's true, maybe it isn't necessary to note anything here.  Its just those of us who know that internally scrub and resilver are nearly the same, we might wonder but we can also just check the code :-)

alek-p commented:
Thanks, @ahrens. I think I've addressed all the comments.
Now working on the port to OpenZFS.

alek-p commented:
during the port to OpenZFS noticed that I didn't update the message for `zpool_scrub_002_pos.ksh` to reflect that it is also testing `scrub -p`. That was the only change in the latest commit.

alek-p commented on man/man8/zpool.8:
```diff
@@ -16,7 +16,7 @@
 .\" own identifying information:
 .\" Portions Copyright [yyyy] [name of copyright owner]
-.TH zpool 8 "April 12, 2017" "ZFS pool 28, filesystem 5" "System Administration Commands"
+.TH zpool 8 "June 16, 2017" "ZFS pool with feature flags, filesystem 5" "System Administration Commands"
```
I've also changed 28 to "with feature flags", hope that's ok

alek-p commented:
The change in the latest commit also modified `zpool_scrub_003_pos.ksh`. Previously it was passing even though it shouldn't.

This is a test that hasn't been valid for over 7 years (since May 2010). It was meant to confirm that issuing a `zpool scrub` when a scrub was already ongoing would restart a scrub. 
However, part of [this ](https://github.com/openzfs/openzfs/commit/3f9d6ad7) commit changed the behavior to not allowing a silent restart of scrubs by issuing `zpool scrub` [again](https://github.com/openzfs/openzfs/commit/3f9d6ad7#diff-1d524334ba4888449557516b778c372aR217): "6943992 'zpool scrub' should not restart the existing scrub silently". Which is how scrub behaves today, it will errors out if you try to issue a scrub when one is ongoing.

I've removed the reference to this behavior from the man page as part of this PR. The new behavior of `zpool_scrub_003_pos.ksh` is to check that issuing a scrub when one is already ongoing does indeed fail.

alek-p commented:
Found out during illumos testing that it will reissue a syscall if the syscall returns ERESTART. In this update, I've switched scrub pause resume to use ECANCELED instead, since I had to do that in the OpenZFS patch.

behlendorf commented on module/zfs/dsl_scan.c:
```diff
@@ -403,6 +406,94 @@ dsl_scan_cancel(dsl_pool_t *dp)
+	dsl_pool_t *dp = dmu_tx_pool(tx);
+	dsl_scan_t *scn = dp->dp_scan;
+
+	ASSERT(*cmd == POOL_SCRUB_NORMAL || *cmd == POOL_SCRUB_PAUSE);
```
**nit**: I know that all the callers verify that `cmd` is valid before calling this so the ASSERT is safe.  However, since the purpose of the `_check()` function as described in `dsl_synctask.c` is to determine exactly this kind of thing it would make sense to return ENOTSUP for an invalid cmd rather that ASSERT.  The check in `spa_scrub_pause_resume()` could then be removed.  Since this isn't ever going to be a hot code path it's not critial that calling this via `dsl_sync_task` is a bit more expensive.

How about

```c
        pool_scrub_cmd_t cmd = *(pool_scrub_cmd_t *)arg;

        if (cmd == POOL_SCRUB_PAUSE) {
                ...
        } else if (cmd == POOL_SCRUB_NORMAL) {
                ...
        } else {
                return (SET_ERROR(ENOTSUP));
        }
```

behlendorf commented:
@alek-p it looks like the [kmemleak builder](http://build.zfsonlinux.org/builders/Ubuntu%2016.04%20x86_64%20Kmemleak%20%28TEST%29/builds/2277) failed the updated test case.

alek-p commented:
I've tried to reproduce the failure without success. Only thing I can think of is that the scrub completed before that check for it running. Then again that's what the zinjected delays are there to prevent.

alek-p commented:
I've added and set a verbose output option for `is_pool_*` functions in `libtest.shlib` so that if the test fails agian we'll have more insight into why.

alek-p commented:
hmm something doesn't look right in that kmemleak tests. It's saying `zpool_scrub_002_pos` test failed but I can find any scrub test outputs in the log

behlendorf commented:
Right, let's try doubling the delay.

behlendorf commented:
Looks good.  @ahrens do you have any additional comments.

ahrens commented:
@behlendorf I'll take another look at this and the OpenZFS version, but it might not be until next week.

behlendorf commented:
@ahrens thanks!  @alek-p unless you have any additional pending changes I'm going to go ahead and merges this.  Let's incorporate any additional feedback in a followup PR if needed.

alek-p commented:
@behlendorf sounds good, I don't have anything else currently
